packages/node_modules/@webex/internal-plugin-conversation/src/conversation.js

```diff
@@ -901,9 +903,96 @@ const Conversation = WebexPlugin.extend({
     this.webex.trigger('user-activity');
 
     return this.request(params)
-      .then((res) => res.body);
+      .then((res) => res.body)
+      .catch((error) => {
+        // handle when key need to rotate
+        if (error.body && error.body.errorCode === KEY_ROTATION_REQUIRED) {
+          params.body.defaultActivityEncryptionKeyUrl = null;
+          params.body.url = params.body.target.url;
+          params.body.objectType = 'conversation';
+          if (activity.object.files.items.length > 0) {
+            this.resendShareActivity(params, activity);
+          }
+          else {
+            this.resendPostActivity(params, activity);
+          }
+        }
+        else if (
+          error.body &&
+    (error.body.errorCode === KEY_ALREADY_ROTATED || error.body.errorCode === ENCRYPTION_KEY_URL_MISMATCH)
+        ) {
+          // handle when key need to update
+          this.webex.request({
+            method: 'GET',
+            api: 'conversation',
+            resource: `conversations/${params.body.target.id}`
+          }).then((res) => {
+            if (activity.object.files.items.length > 0) {
+              this.resendShareActivity(res, activity);
+            }
+            else {
+              this.resendPostActivity(res, activity);
+            }
+          });
+        }
+        else {
+          throw error;
+        }
+      });
   },
+  /**
+   * Resend lost post activity
+   * @param {res} res
+   * @param {activity} activity
+   * @returns {Promise}
+   */
+  resendPostActivity(res, activity) {
+    const postPromise = [];
+    const transform = decryptTransforms.find(({name}) => name === 'decryptTextProp');
+
+    postPromise.push(transform.fn(this, 'content', activity.encryptionKeyUrl, activity.object));
+    postPromise.push(transform.fn(this, 'displayName', activity.encryptionKeyUrl, activity.object));
+
+    Promise.all(postPromise).then(() => {
+      this.post(res.body, activity.object, null);
+    });
+  },
+
+
+  /**
+    * Resend lost share activity
+    * @param {res} res
+    * @param {activity} activity
+    * @returns {Promise}
+    */
+  resendShareActivity(res, activity) {
```

Comment:
@aimeex3 @adamweeks @InteractiveTimmy @taymoork2 - this still feels like an antipattern to me. Dalip said the main source of the issue is that `params` gets modified in place when the request is sent/encrypted. What are your thoughts on cloning the request params before sending an activity? It would do it on _every_ send which isn't great but something about this feels like it will be hard to debug and is going against the typical transform patterns

Reply:
@andrewholsted @chrisadubois @taymoork2 I did comparison using cloneDeep for copying activity object and sending message and without cloneDeep and observed time in ms. I did not see major difference but I considered two cases images and text there might be some scenario when it cause large time difference. Let me know if i need to follow clone activity approach.

|without_cloneDeep|without_cloneDeep|with_cloneDeep|with_cloneDeep|
---|---|---|---|
|text(ms) | image(ms)|text(ms)|image(ms)|
|---|---|---|---|
|296|560|477|602
|278|670|371|554
 271|593|413|585
  288|542|375|613
  283|520|382|627
  292|544|518|587
  306|552|421|590
  412||415
  319||434
  344||456
  320||443
  332||456
  358||463
  405||479
  358||501
  350||525
  378||498
  368||571